src/coreclr/jit/optimizer.cpp

```diff
@@ -5555,14 +5664,25 @@ BasicBlock* Compiler::optInsertLoopChoiceConditions(LoopCloneContext* context,
 
         curCond->inheritWeight(head);
         curCond->bbNatLoopNum = head->bbNatLoopNum;
-        JITDUMP("Created new " FMT_BB " for new level\n", curCond->bbNum);
+        JITDUMP("Created new " FMT_BB " for new level %u\n", curCond->bbNum, i);
     }
 
     // Finally insert cloning conditions after all deref conditions have been inserted.
     context->CondToStmtInBlock(this, *(context->GetConditions(loopNum)), curCond, false);
     return curCond;
 }
 
+//------------------------------------------------------------------------
+// OptEnsureUniqueHead: Ensure that loop "loopInd" has a unique head block.
```

Comment:
It would be nice to see us more standard terminology. But maybe save this for a subsequent PR.

In the literature, this block is usually called a `preheader` (see eg Muchnick section 7.4) and the `header` is the block in the loop that can be entered from outside the loop.

Reply:
Not only the terminology, but also the semantics are weird. What Muchnick calls the header block we call the entry. Our head block is the lexical immediate (non-loop) predecessor of our "first" block. So is what this is creating a pre-header? Sounds like it, but our weird semantics make it hard to understand that.

Reply:
There's also `fgCreateLoopPreHeader()`